src/Dependencies/PooledObjects/ArrayBuilder.cs

```diff
@@ -272,7 +272,16 @@ public ImmutableArray<U> ToDowncastedImmutable<U>()
         /// </summary>
         public ImmutableArray<T> ToImmutableAndFree()
         {
-            var result = this.ToImmutable();
+            ImmutableArray<T> result;
+            if (_builder.Capacity == Count)
+            {
+                result = _builder.MoveToImmutable();
+            }
+            else
+            {
+                result = ToImmutable();
+            }
```

Comment:
Tried this change a long time ago but bcaked off because I had no way of measuring the results. Really glad we have a way to measure the allocaitons now. There are a lot of oppurtunities like this laying around.